### DISCUSSION TITLE
Job is not adhering to the if check

### DISCUSSION THREAD

**Discussion by tsposato** [community/community#24987] [Actions]
<p>Hi,</p>
<p>I have an initial job which checks AWS ECR if an image exists and then sets an output which I use in the next job. This works fine in one of my repositories, I.E when BUILD = ‘NO’ the job is skipped. For some reason in this repository the job always gets ran and I have no idea why. Here is the relevant config.</p>
<pre><code class="lang-auto">jobs:
  Check_For_Existing_Image:
    name: Check if image already exists in ECR
    runs-on: ubuntu-latest
    outputs:
      BUILD: ${{ steps.checkimage.outputs.BUILD }}
    steps:
      - uses: actions/checkout@v2

      - name: Get short SHA
        id: slug
        run: echo "::set-output name=sha7::$(echo ${GITHUB_SHA} | cut -c1-7)"

      - name: Configure AWS Credentials
        uses: aws-actions/configure-aws-credentials@v1
        with:
          aws-access-key-id: ${{ env.AWS_ACCESS_KEY_ID }}
          aws-secret-access-key: ${{ env.AWS_SECRET_ACCESS_KEY }}
          aws-region: ${{ env.AWS_REGION }}

      - name: Check Image
        id: checkimage
        run: |
          EXIT_CODE_NGINX=0
          EXIT_CODE_PHP=0
          aws ecr describe-images --repository-name=${{ env.IMAGENGINX }} --image-ids=imageTag=${{ steps.slug.outputs.sha7 }} 2&gt; /dev/null || EXIT_CODE_NGINX=$?
          aws ecr describe-images --repository-name=${{ env.IMAGEPHPFPM }} --image-ids=imageTag=${{ steps.slug.outputs.sha7 }} 2&gt; /dev/null || EXIT_CODE_PHP=$?

          if [[ "$EXIT_CODE_NGINX" -eq 0 &amp;&amp; "$EXIT_CODE_PHP" -eq 0 ]]; then
              echo "Image ${{ steps.slug.outputs.sha7 }} found"
              echo "::set-output name=BUILD::NO"
          else
              echo "Image ${{ steps.slug.outputs.sha7 }} not found"
              echo "::set-output name=BUILD::YES"
          fi

  Build_Push_Docker_Images:
    name: Build and Push Docker Images
    runs-on: ubuntu-latest
    needs: [ Check_For_Existing_Image ]
    if: ${{ needs.Check_For_Existing_Image.outputs.BUILD }} == 'YES'
    steps:
      - name: Test
        id: test
        run: echo "Value of Build is ${{ needs.Check_For_Existing_Image.outputs.BUILD }}"

      - uses: actions/checkout@v2

      - name: Notify dedicated teams channel
        uses: jdcargile/ms-teams-notification@master
        with:
          github-token: ${{ github.token }} # this will use the runner's token.
          ms-teams-webhook-uri: ${{ env.MSTEAMSWEBHOOK }}
          notification-summary: Build Started
          notification-color: 17a2b8
          timezone: Australia/Victoria

      - name: Get short SHA
        id: slug
        run: echo "::set-output name=sha7::$(echo ${GITHUB_SHA} | cut -c1-7)"

      - name: Cache docker images
        uses: satackey/action-docker-layer-caching@v0.0.11
        with:
          key: ${{ runner.os }}-docker-${{ hashFiles('**/Dockerfile*') }}
          restore-keys: |
            ${{ runner.os }}-docker-
            ${{ runner.os }}-

      - uses: kciter/aws-ecr-action@master
        with:
          access_key_id: ${{ env.AWS_ACCESS_KEY_ID }}
          secret_access_key: ${{ env.AWS_SECRET_ACCESS_KEY }}
          account_id: ${{ env.AWS_ACCOUNT_ID }}
          repo: ${{ env.IMAGEPHPFPM }}
          region: ${{ env.AWS_REGION }}
          tags: latest,${{ steps.slug.outputs.sha7 }}
          dockerfile: ./docker/php-fpm/Dockerfile.prod

      - uses: kciter/aws-ecr-action@master
        with:
          access_key_id: ${{ env.AWS_ACCESS_KEY_ID }}
          secret_access_key: ${{ env.AWS_SECRET_ACCESS_KEY }}
          account_id: ${{ env.AWS_ACCOUNT_ID }}
          repo: ${{ env.IMAGENGINX }}
          region: ${{ env.AWS_REGION }}
          tags: latest,${{ steps.slug.outputs.sha7 }}
          dockerfile: ./docker/nginx/Dockerfile.prod

      - name: Notify dedicated teams channel
        uses: jdcargile/ms-teams-notification@master
        if: success()
        with:
          github-token: ${{ github.token }} # this will use the runner's token.
          ms-teams-webhook-uri: ${{ env.MSTEAMSWEBHOOK }}
          notification-summary: Build Success &amp;#x2705
          notification-color: 65F015
          timezone: Australia/Victoria

      - name: Notify dedicated teams channel
        uses: jdcargile/ms-teams-notification@master
        if: failure()
        with:
          github-token: ${{ github.token }} # this will use the runner's token.
          ms-teams-webhook-uri: ${{ env.MSTEAMSWEBHOOK }}
          notification-summary: Build Failure &amp;#x274C
          notification-color: FB0606
          timezone: Australia/Victoria
</code></pre>
<p>No matter what, the second job runs, when I would assume it shouldn’t.<br>
Here is the relevant debug logs from the Test step in the second job proving the value of BUILD isn’t YES:</p>
<pre><code class="lang-auto">##[debug]/bin/bash -e /home/runner/work/_temp/e59fac71-1656-4e17-b021-9096bcdcfede.sh
Value of Build is NO
##[debug]Finishing: Test
</code></pre>
<p>Is there something I am overlooking here?</p>
<p>Thanks,<br>
Terry</p>

**Comment by airtower-luna** [comment, marked as answer]
<aside class="quote no-group" data-username="tsposato" data-post="1" data-topic="159707">
<div class="title">
<div class="quote-controls"></div>
 tsposato:</div>
<blockquote>
<pre><code class="lang-yaml">    if: ${{ needs.Check_For_Existing_Image.outputs.BUILD }} == 'YES'
</code></pre>
</blockquote>
</aside>
<p>The problem is that you have only <em>part</em> of your conditional wrapped in <code>${{ ... }}</code>. That way only the wrapped part is evaluated as an expression, resulting in a string like <code>NO == 'YES'</code>. And a non-empty string evaluates to <code>true</code>.</p>
<p>For an <code>if</code> it’s easiest to just remove the <code>${{ ... }}</code> part:</p>
<pre><code class="lang-yaml">    if: needs.Check_For_Existing_Image.outputs.BUILD == 'YES'
</code></pre>

**Comment by tsposato** [comment]
<p>Thanks! This did indeed fix my issue.<br>
I had a similar if statement on another job which was always successful for another reason which was throwing me off.</p>
<p>Cheers.</p>
